pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,202 @@
+	// Scan blocks in source directory
+	es, err := os.ReadDir(source)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read directory %q", source)
+	}
+	var failed []string
+	var succeeded []string
+	var alreadyExists []string
+	for _, e := range es {
+		pth := filepath.Join(source, e.Name())
+		if err := c.backfillBlock(ctx, pth, logger); err != nil {
+			if errors.Is(err, errConflict) {
+				level.Warn(logger).Log("msg", "failed uploading block since it already exists on server",
+					"path", pth)
+				alreadyExists = append(alreadyExists, pth)
+			} else {
+				level.Warn(logger).Log("msg", "failed uploading block", "path", pth, "err", err)
+				failed = append(failed, pth)
+			}
+			continue
+		}
+
+		level.Info(logger).Log("msg", "successfully uploaded block", "path", pth)
+		succeeded = append(succeeded, pth)
+	}
+
+	level.Info(logger).Log("msg", "finished uploading block(s)", "succeeded", len(succeeded),
+		"already_exists", len(alreadyExists), "failed", len(failed))
+
+	if len(failed) > 0 {
+		return fmt.Errorf("failed to upload %d block(s)", len(failed))
+	}
+
+	return nil
+}
+
+func closeResp(resp *http.Response) {
+	// Drain and close the body to let the transport reuse the connection
+	io.Copy(io.Discard, resp.Body)
+	resp.Body.Close()
+}
+
+func (c *MimirClient) backfillBlock(ctx context.Context, dpath string, logger log.Logger) error {
+	blockMeta, err := getBlockMeta(dpath)
+	if err != nil {
+		return err
+	}
+
+	blockID := blockMeta.ULID.String()
+
+	logger = log.With(logger, "user", c.id, "block", blockID)
+
+	level.Info(logger).Log("msg", "Making request to start block backfill")
+
+	blockPrefix := path.Join("/api/v1/upload/block", url.PathEscape(blockID))
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(blockMeta); err != nil {
+		return errors.Wrap(err, "failed to JSON encode payload")
+	}
+	resp, err := c.doRequest(blockPrefix, http.MethodPost, buf, int64(buf.Len()))
+	if err != nil {
+		return errors.Wrap(err, "request to start backfill failed")
+	}
+	defer closeResp(resp)
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("request to start backfill failed, with HTTP status %d %s", resp.StatusCode, resp.Status)
+	}
+
+	// Upload each block file
+	if err := filepath.WalkDir(dpath, func(pth string, e fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if e.IsDir() {
+			return nil
+		}
+
+		if filepath.Base(pth) == "meta.json" {
+			// Don't upload meta.json in this step
+			return nil
+		}
+
+		f, err := os.Open(pth)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open %q", pth)
+		}
+		defer f.Close()
+
+		st, err := f.Stat()
+		if err != nil {
+			return errors.Wrap(err, "failed to get file info")
+		}
+
+		relPath, err := filepath.Rel(dpath, pth)
+		if err != nil {
+			return errors.Wrap(err, "failed to get relative path")
+		}
+		relPath = filepath.ToSlash(relPath)
```

Comment:
In `getBlockMeta` we only include selected files in `meta.json`. In this function we upload any file that happens to be in the blocks directory. We should be consistent. (Blocks may have extra files that Mimir doesn't use, eg. Prometheus may create `tombstones` file.)

Perhaps instead of iterating through the directory, we can upload files mentioned in metadata (ignoring meta.json)?